### PR TITLE
map of unequal length bitarray

### LINKED ISSUE #47011: `map!` errors on a `BitVector` when the destination is longer than the source
The docstring for `map!` says "`destination` must be at least as large as the smallest collection." But this fails for `BitVector`s:

```julia
julia> a = falses(3)
3-element BitVector:
 0
 0
 0

julia> map!(!, a, BitVector([0, 1]))
ERROR: DimensionMismatch: sizes of dest and A must match

julia> map!(&, a, BitVector([0, 1]), BitVector([1, 1]))
ERROR: DimensionMismatch: sizes of dest, A, and B must all match
```

### PR DESCRIPTION
**Pull request by Moelf**
fix #47011
- fix bit_map!
- add bit array test for unequal length

### CODE REVIEW

**Comment by CameronBieganek** [comment]
That was fast. 🙂

File: base/bitarray.jl [hunk 1]
```diff
@@ -1779,26 +1779,38 @@ end
 # map across the chunks. Otherwise, fall-back to the AbstractArray method that
 # iterates bit-by-bit.
 function bit_map!(f::F, dest::BitArray, A::BitArray) where F
-    size(A) == size(dest) || throw(DimensionMismatch("sizes of dest and A must match"))
+    size(A) <= size(dest) || throw(DimensionMismatch("size of dest must be >= than size of A"))
```
Comment (N5N3) [inline base/bitarray.jl:1782]:
```suggestion
    length(A) <= length(dest) || throw(DimensionMismatch("length of dest must be >= than length of A"))
```

File: base/bitarray.jl [hunk 2]
```diff
@@ -1779,26 +1779,38 @@ end
...
     isempty(A) && return dest
     destc = dest.chunks
     Ac = A.chunks
-    for i = 1:(length(Ac)-1)
+    len_destc = length(destc)
+    for i = 1:(len_destc-1)
```
Comment (N5N3) [inline base/bitarray.jl:1787]: We should use `len_Ac` here as `len_Ac <= len_destc`

File: base/bitarray.jl [hunk 3]
```diff
@@ -1779,26 +1779,38 @@ end
...
         destc[i] = f(Ac[i])
     end
-    destc[end] = f(Ac[end]) & _msk_end(A)
+    dest_last = destc[len_destc]
+    _msk = _msk_end(A)
+    # first zero out the bits mask is going to change
+    destc[len_destc] = (dest_last & ~(_msk))
+    # then update bits by `or`ing with a masked RHS
+    destc[len_destc] |= f(Ac[len_destc]) & _msk
```
Comment (N5N3) [inline base/bitarray.jl:1795]: Same above. (Would be good to add test for length difference > 64)

File: base/bitarray.jl [hunk 4]
```diff
@@ -1779,26 +1779,40 @@ end
...
+    # first zero out the bits mask is going to change
+    destc[len_Ac] = (dest_last & ~(_msk))
+    # then update bits by `or`ing with a masked RHS
+    destc[len_Ac] |= f(Ac[end]) & _msk
     dest
 end
 function bit_map!(f::F, dest::BitArray, A::BitArray, B::BitArray) where F
-    size(A) == size(B) == size(dest) || throw(DimensionMismatch("sizes of dest, A, and B must all match"))
+    length(A) == length(B) || throw(DimensionMismatch("size of A and B must be equal"))
+    length(A) <= length(A) || throw(DimensionMismatch("size of dest must be >= than size of A and B"))
```
Comment (N5N3) [inline base/bitarray.jl:1801]:
```suggestion
    length(A) <= length(dest) || throw(DimensionMismatch("size of dest must be >= than size of A and B"))
```

File: base/bitarray.jl [hunk 5]
```diff
@@ -1779,26 +1779,40 @@ end
...
     isempty(A) && return dest
     destc = dest.chunks
     Ac = A.chunks
     Bc = B.chunks
-    for i = 1:(length(Ac)-1)
+    for i = 1:(length(destc)-1)
```
Comment (N5N3) [inline base/bitarray.jl:1806]:
```suggestion
    len_Ac = length(Ac)
    for i = 1:len_Ac-1
```

File: base/bitarray.jl [hunk 6]
```diff
@@ -1779,26 +1779,40 @@ end
...
+    _msk = _msk_end(A)
+    # first zero out the bits mask is going to change
+    destc[len_Ac] = (dest_last & ~(_msk))
+    # then update bits by `or`ing with a masked RHS
+    destc[len_Ac] |= f(Ac[end]) & _msk
     dest
 end
 function bit_map!(f::F, dest::BitArray, A::BitArray, B::BitArray) where F
-    size(A) == size(B) == size(dest) || throw(DimensionMismatch("sizes of dest, A, and B must all match"))
+    length(A) == length(B) || throw(DimensionMismatch("size of A and B must be equal"))
```
Comment (N5N3) [inline base/bitarray.jl:1800]: I think we can loose this restriction a bit.
Just set `len_Ac = min(length(Ac), length(BC))` and `_msk = _msk_end(len_Ac)`

Comment (Moelf) [reply base/bitarray.jl:1800]: I don't want people to silently feed the wrong arrays though, at least this is looser than before

Comment (N5N3) [reply base/bitarray.jl:1800]: But that matches the general behavior of `map(!)`.
We can always call `@invoke map!(f, dest::AbstractArray, A::AbstractArray, B::AbstractArray)` in this case.
If `bit_map!` gives the same result with faster speed, why thrown an error?

**Comment by Moelf** [comment]
hmm, the `min(len, len)` only cares about chunk length, we actually don't know what's the true length...........

**Review by N5N3 — APPROVED** [review]

File: base/bitarray.jl [hunk 7]
```diff
@@ -1779,26 +1779,42 @@ end
...
+    # first zero out the bits mask is going to change
+    destc[len_Ac] = (dest_last & ~(_msk))
+    # then update bits by `or`ing with a masked RHS
+    destc[len_Ac] |= f(Ac[end]) & _msk
     dest
 end
 function bit_map!(f::F, dest::BitArray, A::BitArray, B::BitArray) where F
-    size(A) == size(B) == size(dest) || throw(DimensionMismatch("sizes of dest, A, and B must all match"))
+    length(A) <= length(dest) || throw(DimensionMismatch("size of dest must be >= than size of A and B"))
+    length(B) <= length(dest) || throw(DimensionMismatch("size of dest must be >= than size of A and B"))
```
Comment (CameronBieganek) [inline base/bitarray.jl:1801]: The docstring for `map!` says "`destination` must be at least as large as the smallest collection", so `dest` only needs to be `>=` to either `A` or `B`, not both.

File: test/bitarray.jl [hunk 8]
```diff
@@ -1494,6 +1494,37 @@ timesofar("reductions")
...
+                original_tail = last(b1, extra_l)
+                for op in (!, ~)
+                    map!(op, b1, b2)
+                    @test first(b1, l) == map(op, b2)
+                    # check we didn't change bits we're not suppose to
+                    @test last(b1, extra_l) == original_tail
+                end
+
+                b3 = bitrand(l)
+                b4 = bitrand(l+extra_len)
```
Comment (CameronBieganek) [inline test/bitarray.jl:1513]:
```suggestion
                b4 = bitrand(l+extra_l)
```

File: test/bitarray.jl [hunk 9]
```diff
@@ -1494,6 +1494,37 @@ timesofar("reductions")
...
+                    map!(op, b1, b2, b3)
+                    @test first(b1, l) == map(op, b2, b3)
+                    # check we didn't change bits we're not suppose to
+                    @test last(b1, extra_l) == original_tail
+
+                    map!(op, b1, b2, b4)
+                    @test first(b1, l) == map(op, b2, b4)
+                    # check we didn't change bits we're not suppose to
+                    @test last(b1, extra_l) == original_tail
+                end
```
Comment (CameronBieganek) [inline test/bitarray.jl:1524]: Maybe add a test where the length of `dest` is somewhere between the lengths of the two source vectors? Right now it looks like we're not testing that case.

File: base/bitarray.jl [hunk 10]
```diff
@@ -1779,26 +1779,41 @@ end
 # map across the chunks. Otherwise, fall-back to the AbstractArray method that
 # iterates bit-by-bit.
 function bit_map!(f::F, dest::BitArray, A::BitArray) where F
-    size(A) == size(dest) || throw(DimensionMismatch("sizes of dest and A must match"))
+    length(A) <= length(dest) || throw(DimensionMismatch("length of dest must be >= than length of A"))
```
Comment (CameronBieganek) [inline base/bitarray.jl:1782]: `>=` is read "greater than or equal to", so it makes more sense to omit "than" from the sentence. Also, the `map!` docstring names the function arguments `destination` and `collection...`. (There is no docstring specific to the `BitArray` method.)

```suggestion
    length(A) <= length(dest) || throw(DimensionMismatch("length of destination must be >= length of collection"))
```

File: base/bitarray.jl [hunk 11]
```diff
@@ -1779,26 +1779,41 @@ end
...
+    _msk = _msk_end(A)
+    # first zero out the bits mask is going to change
+    destc[len_Ac] = (dest_last & ~(_msk))
+    # then update bits by `or`ing with a masked RHS
+    destc[len_Ac] |= f(Ac[end]) & _msk
     dest
 end
 function bit_map!(f::F, dest::BitArray, A::BitArray, B::BitArray) where F
-    size(A) == size(B) == size(dest) || throw(DimensionMismatch("sizes of dest, A, and B must all match"))
+    min(length(A), length(B)) <= length(dest) || throw(DimensionMismatch("size of dest must be >= than size of A or B"))
```
Comment (CameronBieganek) [inline base/bitarray.jl:1800]: Same as above.

```suggestion
    min(length(A), length(B)) <= length(dest) || throw(DimensionMismatch("size of destination must be >= size of smallest input collection"))
```

**Pull request merged**